File: engine-base/src/main/scala/com/sos/scheduler/engine/base/process/ProcessSignal.java
package com.sos.scheduler.engine.base.process;

import com.sos.scheduler.engine.base.sprayjson.JavaEnumJsonFormat;
import spray.json.JsonFormat;

/**
 * See Unix `kill` command.
 *
 * @author Joacim Zschimmer
 */
public enum ProcessSignal {
    SIGTERM,
    SIGKILL;

    public static final JsonFormat<ProcessSignal> MyJsonFormat = new JavaEnumJsonFormat<>(ProcessSignal.class);
}

File: js7-base/shared/src/main/scala/js7/base/io/process/ProcessSignal.scala
package js7.base.io.process

import io.circe.{Decoder, DecodingFailure, Encoder, Json}

/**
  * @author Joacim Zschimmer
  */
sealed abstract class ProcessSignal(val number: Int)

object ProcessSignal:
  case object SIGTERM extends ProcessSignal(15)
  case object SIGKILL extends ProcessSignal(9)

  implicit val jsonEncoder: Encoder[ProcessSignal] =
    o => Json.fromString(o.toString)

  implicit val jsonDecoder: Decoder[ProcessSignal] =
    cursor => cursor.as[String].flatMap:
      case "SIGTERM" => Right(SIGTERM)
      case "SIGKILL" => Right(SIGKILL)
      case signal => Left(DecodingFailure(s"Invalid process signal: $signal (SIGTERM or SIGKILL expected)", cursor.history))
